File: haskell/grains/src/Grains.hs
module Grains (square, total) where

import Data.Maybe (fromMaybe)

square :: Integer -> Maybe Integer
square n
  | n <= 0 = Nothing
  | n >= 65 = Nothing
  | otherwise = Just (2 ^ (n - 1))

total :: Integer
total = fromMaybe 0 (square 64) * 2 - 1

File: erlang/grains/src/grains.erl
-module(grains).
-export([square/1, total/0]).

square(Square) when Square < 1; Square > 64 -> {error, "square must be between 1 and 64"};
square(Square) -> 1 bsl (Square - 1).

total() ->
  % square(65) - 1.
  18446744073709551615.
